### PR TITLE
[self-guided] tour pages

### PR DESCRIPTION
**Pull request by dvdherron**
## Cute animal pic
image

_Because everyone likes pictures of [animals](https://unsplash.com/t/animals)._

## Link to Trello card (if applicable)
https://trello.com/c/8oJo98Ex/261-as-a-new-user-i-want-to-view-more-information-about-the-metecho-ui-in-a-nonlinear-fashion-self-guided-tour-on-non-project-detail
_If this PR relates to a Trello card, don't forget to reference this PR on the card as well._

## Description
_The commit messages say what you did; this should explain why and/or how._
- [x ] Description is in Trello card

## Steps to test/reproduce

- Turn on the self-guided tour.
- Look for and open the tour popovers on the following pages
  - Project List
  - Epic Detail
  - Task Detail
_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

### Project List
image

### Epic Detail
image

### Task Detail
image

_Provide screenshots/animated gifs/videos if necessary._

### CODE REVIEW

**Comment by dvdherron** [comment]
@jgerigmeyer These are mostly in place. I think it can be ready for a review soon?

There are two that are missing: one for the Commit History and the one for the breadcrumb if a Task is not grouped in an Epic. 

- How do I get the commit history component to render?
- For the breadcrumb, do you see a way to better way to place the popovers there in order to have them appear over specific items in that list?

There's sort of a similar placement issue on the project list page. I have it on the left side of the entire list for now unless there's a way to put it only on the first item in that whole list.

File: src/js/components/epics/detail.tsx
Comment (jgerigmeyer) [inline src/js/components/epics/detail.tsx:416]: Both here (epic detail) and on task detail, we want this popover to change depending whether the button is "View Branch", "View Changes", or "View Pull Request". You can see the logic for this in `helpers.ts`, `getBranchLink` fn. Maybe we want a similar fn to get the right popover? Or we could repurpose `getBranchLink` to also return the popover?

**Review by jgerigmeyer — COMMENTED** [review]
@dvdherron This is looking good! I didn't do a comprehensive code review, but I did functional review and tried to answer your questions.

- For the commit history, if there are commits on the relevant GH branch (you can make these manually in GH outside Metecho if that's easier), you can run `derrick manage resync_all_gh_commit_data` or (`python manage.py resync_all_gh_commit_data` inside a container terminal) to get the list of commits to show up in Metecho.
- For the breadcrumb, skip the "no epic" task breadcrumb for now. That's not something we currently support.

If any of my comments don't make sense, I'm happy to jump on a call and screenshare in more detail!

**Comment by dvdherron** [comment]
> 
> 
> @dvdherron This is looking good! I didn't do a comprehensive code review, but I did functional review and tried to answer your questions.
> 
>     * For the commit history, if there are commits on the relevant GH branch (you can make these manually in GH outside Metecho if that's easier), you can run `derrick manage resync_all_gh_commit_data` or (`python manage.py resync_all_gh_commit_data` inside a container terminal) to get the list of commits to show up in Metecho.
> 
>     * For the breadcrumb, skip the "no epic" task breadcrumb for now. That's not something we currently support.
> 
> 
> If any of my comments don't make sense, I'm happy to jump on a call and screenshare in more detail!

Fantastic. Thanks for the feedback! Working through these issues now and I'll circle back if I run into any trouble.

File: src/js/components/tour/popover.tsx
Comment (dvdherron) [reply src/js/components/tour/popover.tsx]: @jgerigmeyer There are no popovers without headings but without this the popover in `orgActions` causes an error. I'm [changing the header](https://github.com/oddbird/Metecho/pull/1088/files#diff-961f0674b32dc07637039a492502239afd654e6619acffc746c551632ba92f75R173) within the switch statement. Is there a better way to do that?

File: src/js/components/utils/detailPageLayout.tsx
Comment (dvdherron) [reply src/js/components/utils/detailPageLayout.tsx]: Adjusted this in a way that adds the respective popover for the epic detail and task detail pages. had to remove the "slds-truncate" class to get the popover to show. so far I don't see that affecting anything but if you know why that might be an issue I can rethink this.

File: src/js/components/tour/popover.tsx
Comment (jgerigmeyer) [reply src/js/components/tour/popover.tsx]: Ah, yeah -- I see, that makes sense. I think we can remove this, and I'll comment on that other section of code with a suggestion.

File: src/js/components/orgs/cards/orgActions.tsx
Comment (jgerigmeyer) [inline src/js/components/orgs/cards/orgActions.tsx:151]: It might be cleaner to define `let popover = null;` above, then do something like:

```suggestion
        popover = (
          <TourPopover
            align="right"
            heading={i18n.t('Create a Dev Org')}
            body={(
              <Trans i18nKey="tourTaskCreateDevOrg">
                A Dev Org is a temporary Salesforce org where you can make changes that you would like to contribute to the Project. To create an Org, make sure you are connected to your Salesforce account with a Dev Hub enabled. Use the drop down menu to delete the Org when you no longer need it.
              </Trans>
            )}
          />
        );
```

Then you can define `popover` differently for a QA org, and use it below with just `{popover}` after the button.

File: src/js/components/tour/popover.tsx
Comment (jgerigmeyer) [inline src/js/components/tour/popover.tsx:15]:
```suggestion
  heading: string;
```

File: src/js/components/utils/detailPageLayout.tsx
Comment (jgerigmeyer) [inline src/js/components/utils/detailPageLayout.tsx:98]: I think it would work to just always include `{crumb.tourPopover}` and remove this conditional; if it's `undefined` then it won't render anything anyway.

Also, not a big deal but for consistency I'd say we might as well include `{crumb.tourPopover}` in the above `<Link>` version too.

File: src/js/components/utils/detailPageLayout.tsx
Comment (jgerigmeyer) [reply src/js/components/utils/detailPageLayout.tsx]: Hm, I'd want to involve @stacyk or @SondraE in that discussion. It'll mean that longer names will wrap on narrower screens, which also causes some issues with the popover placement.

Screen Shot 2021-05-18 at 12 14 06 PM

Comment (stacyk) [reply src/js/components/utils/detailPageLayout.tsx]: @dvdherron @jgerigmeyer is it possible to wrap only the last child in the breadcrumb and place the popover icon at the beginning of the breadcrumb label it is documenting? Or maybe no need to wrap, but still keep the popover before?

File: src/js/components/epics/detail.tsx
Comment (dvdherron) [reply src/js/components/epics/detail.tsx:394]: added
image

File: src/js/components/tasks/detail.tsx
Comment (dvdherron) [reply src/js/components/tasks/detail.tsx:677]: Added these: 

capture: 
image

submitButton
image

File: src/js/components/projects/list.tsx
Comment (dvdherron) [reply src/js/components/projects/list.tsx]: That worked 👍🏽 
image

File: src/js/components/epics/detail.tsx
Comment (dvdherron) [reply src/js/components/epics/detail.tsx]: A little stuck here. Having a hard time figuring out the syntax to test whether the object is an epic or a task. Is what I'm doing here so far even close to being on the right track?

Comment (jgerigmeyer) [reply src/js/components/epics/detail.tsx]: @dvdherron Yeah, I agree this is a bit tricky, but what you have is definitely on the right track. There would be a number of different ways to approach this, but I'd say the clearest might be adding a second argument that `getBranchLink` accepts: `type: 'epic' | 'task'`. Then I think in the fn body, you could do e.g. `taskPopoverHeading = type === OBJECT_TYPES.EPIC ? i18n.t('View GitHub Branch for Epic') : i18n.t('View GitHub Branch for Task');` -- or something along those lines?

Happy to pair with you if that helps. I'd also say if you get the logic working but are struggling to make TypeScript warnings happy, don't spend much time on that. Just commit and I can look closer at the TS.

Comment (dvdherron) [reply src/js/components/epics/detail.tsx]: Thanks. My meeting went longer than planned. Digging into this now.

Comment (dvdherron) [reply src/js/components/epics/detail.tsx]: was able to get this to work. will give it another look in the AM to see if there's a better way @jgerigmeyer

File: src/js/components/utils/detailPageLayout.tsx
Comment (dvdherron) [reply src/js/components/utils/detailPageLayout.tsx]: @stacyk Yes, that's possible. Something like this @stacyk 

image

 As of right now, that would just require a change in positioning in the CSS. 

The tricky part is making sure the popover is visible with the ` slds-truncate` class added back to the div wrapping the `Breadcrumb` component. The `overflow: hidden` setting cuts off the popover contents:

image

File: src/js/components/orgs/cards/orgActions.tsx
Comment (jgerigmeyer) [inline src/js/components/orgs/cards/orgActions.tsx:203]: This is pretty minor, but it looks like when both of these buttons are visible, the "Create Org" btn appears on top of the "Submit Review" popover icon:

Screen Shot 2021-05-24 at 12 05 13 PM

File: src/js/components/utils/detailPageLayout.tsx
Comment (jgerigmeyer) [inline src/js/components/utils/detailPageLayout.tsx:76]: Just a reminder that this breadcrumb styling is not finalized. I know mobile is a low priority for Metecho, but things get pretty weird on narrow screens...

Screen Shot 2021-05-24 at 12 06 23 PM

**Review by jgerigmeyer — COMMENTED** [review]
@dvdherron This is looking good! I'll work on backfilling JS tests.

File: src/js/components/projects/list.tsx
Comment (dvdherron) [reply src/js/components/projects/list.tsx]: adjusted this. 
image

And I think these two were placed in a similar way. Adjusted these also. 
image

File: src/js/components/orgs/cards/orgActions.tsx
Comment (dvdherron) [reply src/js/components/orgs/cards/orgActions.tsx:203]: I'll work on adding this space when these appear together.

File: src/js/components/utils/detailPageLayout.tsx
Comment (dvdherron) [reply src/js/components/utils/detailPageLayout.tsx:76]: And revisit this

File: src/js/components/orgs/cards/orgActions.tsx
Comment (jgerigmeyer) [reply src/js/components/orgs/cards/orgActions.tsx:203]: Ok -- if it's not simple then we can ignore it; it's hardly noticeable.

File: src/js/components/epics/progress.tsx
Comment (stacyk) [inline src/js/components/epics/progress.tsx:40]: I don't know the best place to leave this @dvdherron, but I see tis progress bar description is also showing up where the progress path description should be. 
task-progress-path

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:40]: zindex 

If we bump this z-index to at least 1, it won't lay behind this table cell neighbor I believe

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:75]: @dvdherron There are quite a few "magic numbers" in this file and I wonder if there is a more systematic way of positioning these? Maybe not, but it may be helpful to see some custom properties or variables https://www.lightningdesignsystem.com/design-tokens/#category-spacing

**Review by stacyk — COMMENTED** [review]
@dvdherron I didn't quite make it through the breadcrumbs, but we can chat after standup. I left a few comments in here, this is looking super good. I noticed some of the text is different from the Self Guided Text file. I don't know if that is a problem or not, but I thought I would mention it.

File: src/js/components/epics/progress.tsx
Comment (dvdherron) [reply src/js/components/epics/progress.tsx:40]: Good catch. Both elements had the same translation key. Fixed. 
image

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:40]: fixed

File: src/js/components/orgs/cards/orgActions.tsx
Comment (dvdherron) [reply src/js/components/orgs/cards/orgActions.tsx:203]: Fixing the z-index helped here. re-positioned the popovers a bit too. 
image

**Comment by jgerigmeyer** [comment]
> I noticed some of the text is different from the Self Guided Text file. I don't know if that is a problem or not, but I thought I would mention it.

@stacyk I did change some of the text intentionally during my review. Mostly three types:

- Where I found the specific wording to be confusing
- Correcting inaccuracies
- Grammatical consistency (e.g. `and` vs `&` in body text vs headings)

If there are specific copy changes that anyone questions, I'm happy to revisit!

File: src/sass/components/_tour.scss
Comment (jgerigmeyer) [inline src/sass/components/_tour.scss:95]: Similar to @stacyk's other comment, but I think we could use `$z-index-default` here: https://www.lightningdesignsystem.com/design-tokens/#category-z-index

Comment (dvdherron) [reply src/sass/components/_tour.scss]: fixed

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss]: @stacyk Made an attempt to clean these up with tokens. Since they are all being contained in different ways, it was tough to come up with a system that would work across the board. Let me know if this helps or not.

File: src/js/components/utils/detailPageLayout.tsx
Comment (dvdherron) [reply src/js/components/utils/detailPageLayout.tsx:111]: What we discussed worked for getting the breadcrumb popover to show and to apply the `slds-truncate `styles to the navigation list items. 
image

@stacyk @jgerigmeyer 
I think this is looking better. The list items aren't wrapping and the popover for the page title moves to the left on narrow screens.

File: src/js/components/utils/detailPageLayout.tsx
Comment (dvdherron) [inline src/js/components/utils/detailPageLayout.tsx:48]: @jgerigmeyer Moved the breadcrumb popover here. Why isn't the text changing? Am I referencing `OBJECT_TYPES` incorrectly?

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:55]: @jgerigmeyer This app is being translated into other languages, right?  If so, then we won't be able to position these popovers based on where the current English text ends up, like this one. 

@dvdherron I think for block level elements like headings, we probably want to position the popover target before the content.

File: src/js/components/commits/list.tsx
Comment (stacyk) [inline src/js/components/commits/list.tsx:109]: @mirisuzanne looking for a second opinion here, but I think section heading elements only accept phrasing content, not divs. Is that correct? 

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements

image

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:92]: Do you know @jgerigmeyer if we ever did truncate each breadcrumb? I looked at staging and we don't do it there. 

@dvdherron One option (new story I assume) I see SLDS has an overflow option https://www.lightningdesignsystem.com/components/breadcrumbs/#With-Overflow 

Another idea **If we want to truncate each `<li>` separately**, which I think is better usability, we would need to do something like 
- updating this line to `.metecho-breadcrumb li` instead of `ol`
- overriding the `<a>` to not have `display: block` 
- optional:  `.metecho-breadcrumb li:not(:first-of-type)` so it doesn't truncate home.
- we'd need to add a way to target the last item, which isn't a link, to tell that div in that last li not to display block... `.metecho-breadcrumb .slds-breadcrumb__item:last-of-type > div { display: inline (or contents) }` ... that is unless we can change it to a span?

Comment (jgerigmeyer) [reply src/sass/components/_tour.scss:92]: > Do you know @jgerigmeyer if we ever did truncate each breadcrumb? I looked at staging and we don't do it there.

I'm not sure -- I'm guessing we didn't?

File: src/sass/components/_tour.scss
Comment (jgerigmeyer) [reply src/sass/components/_tour.scss:55]: Yes, I think that's correct. We shouldn't do any positioning strictly based on the length of text (in current font and language).

File: src/js/components/utils/detailPageLayout.tsx
Comment (jgerigmeyer) [reply src/js/components/utils/detailPageLayout.tsx]: You were _accepting_ a `type` prop here, but not passing one in from where it's used. Fixed in https://github.com/oddbird/Metecho/commit/56cfb220d46b81aade5d2123aebe659e9a51d7da

Comment (dvdherron) [reply src/js/components/utils/detailPageLayout.tsx]: Aha. Got it. Thanks!

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:55]: Makes sense. For consistency's sake then we should probably have all of the popovers placed on the same general side of the elements they are referencing, right? @stacyk 

I'm working on removing them from inside of heading elements based on your comment below.

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:92]: I tried truncating the `<li>`'s separately yesterday but having `overflow-hidden` cut of parts of the list markers and made them all really small @stacyk. I'll play around with this some more in a bit after I reposition the popovers

**Comment by dvdherron** [comment]
@jgerigmeyer @stacyk 
Relocated these to the left-hand side of the page. The task progress bar popover is the last one that needs to be placed. 

image

I'll get to this in the AM.

**Comment by dvdherron** [comment]
@jgerigmeyer I think it's a Python test coverage that's failing? I see 100% on all of the JS tests locally. 

@stacyk we addressed the breadcrumb issue yesterday. I think I addressed the rest of the issues above also if you want to give this another review.

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:74]: @dvdherron this can likely be removed if the suggestion for logo-popover works.

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:50]: @dvdherron If we target `.logo-popover > .popover-wrapper` then we don't need to also include the override for `.user-info &`

Also, these 2 properties can be re-ordered by alpha.

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:79]: @dvdherron same here, we could target `.page-title > .popover-wrapper` and then we probably won't need this. Although, I notice we are using a `-1rem` top value as opposed the the `.75rem` that other buttons use. Is there a reason for the `.25` difference?

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:69]: @dvdherron we can try `.page-title > .popover-wrapper` to only target the direct children of the page-title container

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:97]: @dvdherron I think this is default, and not needed here? If it is needed, then we will want to reorder properties but I think we can remove the left: ...

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:84]: @dvdherron This is another button that has a slightly different positioning from some of the other buttons. If there is a reason to keep this one different than the others, maybe we can document why there are some that benefit from the position overrides. If this isn't needed, we will want to remove it along with the place it is used in the list.tsx file since I think that is the only place it is used.

File: src/js/components/epics/detail.tsx
Comment (stacyk) [inline src/js/components/epics/detail.tsx:574]: @dvdherron are we using the right-edge class for anything?

File: src/js/components/orgs/taskOrgCard.tsx
Comment (stacyk) [inline src/js/components/orgs/taskOrgCard.tsx:224]: @dvdherron if we aren't using right-edge anywhere (currently on the task and epic status sidebars I think) then we could use something on the task cards to push them a little closer to their corner's edge for consistency? Although, we may want a different classname than `right-edge` since that doesn't seem to indicate what it is doing.

File: src/js/components/orgs/cards/orgActions.tsx
Comment (stacyk) [inline src/js/components/orgs/cards/orgActions.tsx:205]: @dvdherron if we add `inline-container` here, then do we need the popover override for `task-create-btn`?

**Review by stacyk — COMMENTED** [review]
@dvdherron this is looking really good. I left some suggestions about some ways that I think could make things more consistent/simplified but you will want to test them out on your end to see if they make sense in context. I didn't have all the scenarios so some of my suggestions may not take certain states into account.

File: src/js/components/commits/list.tsx
Comment (jgerigmeyer) [inline src/js/components/commits/list.tsx:102]: @dvdherron Maybe this is over-engineering, but I wonder if it would be cleaner for the `<TourPopover>` component itself to render its own `<div className="slds-is-relative">` container, and to also accept an optional `className` prop to add e.g. `className="heading"` or whatever. It would sure make the diff simpler from the places where we're adding in popovers... but I don't know if a single wrapping `div` would work in every case?

Comment (jgerigmeyer) [reply src/js/components/commits/list.tsx:102]: Oh, I guess that wouldn't really work because the popover _and_ some arbitrary contents need to be wrapped together... Probably simpler to just leave it as-is then 👍

**Review by jgerigmeyer — COMMENTED** [review]
@dvdherron This is looking really good!

File: src/js/components/epics/progress.tsx
Comment (dvdherron) [reply src/js/components/epics/progress.tsx:17]: yes, this is extra. removed.

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:74]: removed

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:50]: done. didn't need that override there.

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:79]: @stacyk I had originally thought it looked better to get all three of these on a similar level but I think the slight difference on narrow screens is actually fine. 

image

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:96]: yes. the `left` property was redundant. removed.

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss:84]: this was something I missed from the transition from right to left placement. removed.

File: src/js/components/epics/detail.tsx
Comment (dvdherron) [reply src/js/components/epics/detail.tsx:573]: same as above. removed.

File: src/js/components/orgs/taskOrgCard.tsx
Comment (dvdherron) [reply src/js/components/orgs/taskOrgCard.tsx]: made an `.org-card` class and moved them closer

image

File: src/js/components/orgs/cards/orgActions.tsx
Comment (dvdherron) [reply src/js/components/orgs/cards/orgActions.tsx:205]: adding `inline-container` worked. removed `task-create-btn`

File: src/js/components/commits/list.tsx
Comment (dvdherron) [reply src/js/components/commits/list.tsx:102]: yeah, inserting it to each respective reference element probably has to be done manually with this type of popover . . .

File: src/js/components/orgs/taskOrgCard.tsx
Comment (jgerigmeyer) [inline src/js/components/orgs/taskOrgCard.tsx:212]:
```suggestion
        slds-is-relative
        org-card"
```

**Review by jgerigmeyer — APPROVED** [review]

File: src/sass/components/_tour.scss
Comment (stacyk) [reply src/sass/components/_tour.scss]: me too @dvdherron

File: src/sass/components/_tour.scss
Comment (stacyk) [inline src/sass/components/_tour.scss:121]: @dvdherron I'm not sure what happened to my comment on this one so forgive me if it is duplicate, but I think this could move to one of the display (data-helpers or containers) scss files in the patterns folder. It feels like a more generic pattern that can be used elsewhere even though we are currently only using it on tour items at this time.

**Review by stacyk — APPROVED** [review]
@dvdherron One more tiny change I am recommending but I am going to approve this since it is just about organization and not about the code itself :) Great job on this, it is a big branch/task!

File: src/sass/components/_tour.scss
Comment (dvdherron) [reply src/sass/components/_tour.scss]: 👍🏽  moved to `_containers`

**Comment by dvdherron** [comment]
> 
> 
> @dvdherron One more tiny change I am recommending but I am going to approve this since it is just about organization and not about the code itself :) Great job on this, it is a big branch/task!

Thanks @stacyk! Made that last organization change so I think this is ready to merge @jgerigmeyer.

**Review by jgerigmeyer — APPROVED** [review]

**Pull request merged**
